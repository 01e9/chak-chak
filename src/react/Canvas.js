import React from 'react'
import PropTypes from 'prop-types'
import { canvasDrawImage } from "@/utility/canvas";
import { withStyles } from '@material-ui/core'

class Canvas_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired,
        image: PropTypes.instanceOf(Image).isRequired
    }

    canvasRef = React.createRef()

    componentDidMount() {
        canvasDrawImage(this.canvasRef.current, this.props.image);
    }

    render() {
        const { classes } = this.props;
        return (
            <canvas ref={this.canvasRef} className={classes.root} />
        );
    }
}

const styles = {
    root: {
        display: 'block'
    }
}

const Canvas = withStyles(styles)(Canvas_);

export default Canvas;
