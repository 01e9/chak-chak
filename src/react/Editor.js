import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core'

class Editor_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired,
        Canvas: PropTypes.func.isRequired,
        Toolbar: PropTypes.func.isRequired,
        ActiveTool: PropTypes.func.isRequired
    }

    render() {
        const { Canvas, Toolbar, ActiveTool, classes } = this.props;

        return (
            <React.Fragment>
                <div className={classes.canvasWrap}>
                    <Canvas/>
                    <ActiveTool/>
                </div>
                <Toolbar/>
            </React.Fragment>
        )
    }
}

const styles = {
    canvasWrap: {
        position: 'relative'
    }
}

const Editor = withStyles(styles)(Editor_);

export default Editor;
