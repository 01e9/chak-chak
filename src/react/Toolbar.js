import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core'
import Paper from '@material-ui/core/Paper';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import {
    DragIndicator as DragIndicatorIcon,
    Brush as BrushIcon,
    Crop as CropIcon,
    CloudUpload as CloudUploadIcon
} from '@material-ui/icons';

class Toolbar_ extends React.Component {
    static propTypes = {
        classes: PropTypes.object.isRequired
    }

    render() {
        const { classes } = this.props;
        const tabProps = {
            classes: {root: classes.tab}
        }

        return (
            <Paper className={classes.paper}>
                <Tabs>
                    <Tab icon={<DragIndicatorIcon/>} {...tabProps} />
                    <Tab icon={<BrushIcon/>} {...tabProps} />
                    <Tab icon={<CropIcon/>} {...tabProps} />
                    <Tab icon={<CloudUploadIcon/>} {...tabProps} />
                </Tabs>
            </Paper>
        )
    }
}

const styles = {
    paper: {
        margin: '100px',
        display: 'inline-block'
    },
    tab: {
        minWidth: '50px'
    }
}

const Toolbar = withStyles(styles)(Toolbar_)

export default Toolbar
