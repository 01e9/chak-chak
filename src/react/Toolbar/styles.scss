.root {
    position: fixed;
}

.drag {
    cursor: move;
    touch-action: none;
}
