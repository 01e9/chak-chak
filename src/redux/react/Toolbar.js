import * as React from 'react'
import { connect } from 'react-redux'
import Toolbar_ from "@/react/Toolbar";
import { selectToolbarPosition } from "@/redux/selectors/toolbar";
import { actionToolbarPositionSet } from "@/redux/actions/toolbar";
import ToolBrushControl from "@/redux/react/tool/brush/Control";
import { TOOL_NAME_BRUSH } from "@/constants/tools";

const mapStateToProps = (state, ownProps) => {
    const position = selectToolbarPosition(state);

    return {
        left: position.x,
        top: position.y,
        tools: [
            () => <ToolBrushControl key={TOOL_NAME_BRUSH} />
        ]
    }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
    onPositionUpdate: (left, top) => dispatch(actionToolbarPositionSet(left, top))
})

const Toolbar = connect(mapStateToProps, mapDispatchToProps)(Toolbar_)

export default Toolbar
