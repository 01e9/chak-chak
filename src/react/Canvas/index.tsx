import * as React from 'react'
import { canvasDrawImage } from "@/utility/canvas";
import * as styles from "./styles.scss";

export interface ICanvasProps {
    image?: HTMLImageElement;
    aspectRatio: number;
}

export default class Canvas extends React.Component<ICanvasProps> {
    canvasRef = React.createRef<HTMLCanvasElement>()

    redrawImage() {
        if (this.canvasRef.current && this.props.image) {
            canvasDrawImage(this.canvasRef.current, this.props.image, this.props.aspectRatio);
        }
    }

    componentDidMount() {
        this.redrawImage();
    }

    componentDidUpdate() {
        this.redrawImage();
    }

    render() {
        return (
            <canvas ref={this.canvasRef} className={styles.root} />
        );
    }
}
