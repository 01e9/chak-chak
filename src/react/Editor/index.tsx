import * as React from 'react'
import * as styles from "./styles.scss";

export interface IEditorProps {
    Canvas: React.JSXElementConstructor<{}>,
    Toolbar: React.JSXElementConstructor<{}>,
    ActiveTool: React.JSXElementConstructor<{}>
}

class Editor extends React.Component<IEditorProps> {
    render() {
        const { Canvas, Toolbar, ActiveTool } = this.props;

        return (
            <React.Fragment>
                <div className={styles.canvasWrap}>
                    <Canvas/>
                    <ActiveTool/>
                </div>
                <Toolbar/>
            </React.Fragment>
        )
    }
}

export default Editor;
