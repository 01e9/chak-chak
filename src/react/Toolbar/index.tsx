import * as React from 'react'
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import { DragIndicator as DragIndicatorIcon } from '@material-ui/icons';
import * as styles from "./styles.scss";

export interface IToolbarProps {
    left: number;
    top: number;
    onPositionUpdate: (left: number, top: number) => void;
    tools: Array<{key: string; Component: React.JSXElementConstructor<{}>}>;
}

interface IState {
    hasCapture: boolean;
    left: number;
    top: number;
    leftShift: number;
    topShift: number;
}

class Toolbar extends React.Component<IToolbarProps, IState> {
    state: IState = {
        hasCapture: false,
        left: 0,
        top: 0,
        leftShift: 0,
        topShift: 0
    }

    static getDerivedStateFromProps(props: IToolbarProps, state: IState) {
        if (state.hasCapture) {
            return null;
        } else {
            return {
                left: props.left,
                top: props.top
            }
        }
    }

    getPointerPosition(event: React.PointerEvent) {
        return {
            left: event.pageX,
            top: event.pageY
        }
    }

    onDown = (event: React.PointerEvent) => {
        // @ts-ignore setPointerCapture is not defined
        event.target.setPointerCapture(event.pointerId);
    }

    onGotCapture = (event: React.PointerEvent) => {
        const pointerPosition = this.getPointerPosition(event);

        this.setState(({left, top}) => ({
            hasCapture: true,
            leftShift: left - pointerPosition.left,
            topShift: top - pointerPosition.top
        }));
    }

    onMove = (event: React.PointerEvent) => {
        event.stopPropagation();
        const pointerPosition = this.getPointerPosition(event);

        this.setState(({leftShift, topShift}) => ({
            left: pointerPosition.left + leftShift,
            top: pointerPosition.top + topShift
        }));
    }

    onLostCapture = () => {
        this.setState({
            hasCapture: false,
            leftShift: 0,
            topShift: 0
        });

        this.props.onPositionUpdate(
            Math.max(0, this.state.left),
            Math.max(0, this.state.top)
        );
    }

    render() {
        const
            { tools } = this.props,
            { top, left, hasCapture } = this.state;

        return (
            <Paper className={styles.root} style={{top: top + 'px', left: left + 'px'}}>
                <Button
                    className={styles.drag + ' drag-button'}
                    disableRipple

                    onPointerDown={this.onDown}
                    onGotPointerCapture={this.onGotCapture}
                    onPointerMove={hasCapture ? this.onMove : undefined}
                    onLostPointerCapture={this.onLostCapture}
                >
                    <DragIndicatorIcon />
                </Button>
                {tools.map(({key, Component}) => <Component key={key} />)}
            </Paper>
        )
    }
}

export default Toolbar;
