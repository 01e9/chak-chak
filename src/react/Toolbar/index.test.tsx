import * as React from 'react'
import Toolbar, { IToolbarProps } from "@/react/Toolbar";
import { createMount, createShallow } from '@material-ui/core/test-utils';

const requiredProps: IToolbarProps = {
    left: 101,
    top: 102,
    onPositionUpdate: jest.fn(),
    tools: [
        {key: "1", Component: () => <button key="tool1">[Tool1]</button>},
        {key: "2", Component: () => <button key="tool2">[Tool2]</button>},
    ]
}

it('snapshot', () => {
    const wrapper = createShallow()(<Toolbar {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
})

it('moves', () => {
    const props: IToolbarProps = {
        ...requiredProps,
        onPositionUpdate: jest.fn()
    }
    const wrapper = createMount()(<Toolbar {...props} />);

    expect(props.onPositionUpdate).not.toHaveBeenCalled();

    const instance = wrapper.find(Toolbar).instance();
    expect(instance).toBeTruthy();
    expect(instance.state.hasCapture).toBe(false);

    const dragButton = wrapper.find('.drag-button').hostNodes();
    expect(dragButton).toHaveLength(1);

    dragButton.simulate('pointerMove', null);
    expect(instance.state).toMatchObject({
        hasCapture: false
    });

    const
        leftShift = 1,
        topShift = 2,
        leftMove = 10,
        topMove = 20;

    {
        const event = {pointerId: 'test', target: {setPointerCapture: jest.fn()}}
        dragButton.simulate('pointerDown', event);
        expect(event.target.setPointerCapture).toHaveBeenCalledWith('test');
        expect(instance.state.hasCapture).toBe(false);
    }
    {
        const event = {
            pageX: requiredProps.left + leftShift,
            pageY: requiredProps.top + topShift
        }
        dragButton.simulate('gotPointerCapture', event);
        expect(instance.state).toMatchObject({
            hasCapture: true,
            leftShift: -leftShift,
            topShift: -topShift
        });
    }
    {
        const event = {
            pageX: requiredProps.left + leftMove,
            pageY: requiredProps.top + topMove
        }
        dragButton.simulate('pointerMove', event);
        expect(instance.state).toMatchObject({
            hasCapture: true,
            leftShift: -leftShift,
            topShift: -topShift,
            left: requiredProps.left + leftMove - leftShift,
            top: requiredProps.top + topMove - topShift,
        });
    }
    {
        dragButton.simulate('lostPointerCapture', {});
        expect(instance.state).toMatchObject({
            hasCapture: false,
            leftShift: 0,
            topShift: 0
        });
    }

    expect(props.onPositionUpdate).toHaveBeenCalledTimes(1);
    expect(props.onPositionUpdate).toHaveBeenCalledWith(
        requiredProps.left + leftMove - leftShift,
        requiredProps.top + topMove - topShift
    );
})
