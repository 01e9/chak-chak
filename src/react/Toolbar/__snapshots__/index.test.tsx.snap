// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<WithStyles(ForwardRef(Paper))
  style={
    Object {
      "left": "101px",
      "top": "102px",
    }
  }
>
  <WithStyles(ForwardRef(Button))
    className="undefined drag-button"
    disableRipple={true}
    onGotPointerCapture={[Function]}
    onLostPointerCapture={[Function]}
    onPointerDown={[Function]}
  >
    <DragIndicatorIcon />
  </WithStyles(ForwardRef(Button))>
  <Component
    key="1"
  />
  <Component
    key="2"
  />
</WithStyles(ForwardRef(Paper))>
`;
